File: Algorithms/Python/0797_All_Paths_From_Source_to_Target/All_Paths_From_Source_to_Target_Solution_1.py

# Space: O(n)
# Time: O(n)

# BFS solution

class Solution:
    def allPathsSourceTarget(self, graph):
        res = []
        queue = [[[0], graph[0]]]

        while queue:
            path, cur_node_list = queue.pop(0)
            if len(cur_node_list) == 0:
                res.append(path)
            for i in cur_node_list:
                queue.append([path + [i], graph[i]])

        return res

File: Algorithms/0797_All_Paths_From_Source_to_Target/Java/AllPathsFromSourcetoTargetSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    private List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        List<Integer> temp = new ArrayList<>();
        temp.add(0);
        dfs(graph, graph[0], temp, graph.length - 1);
        return res;
    }

    private void dfs(int[][] graph, int[] nodeList, List<Integer> temp, int endNode) {
        if (temp.get(temp.size() - 1) == endNode) {
            res.add(new ArrayList<>(temp));
            return;
        }
        for (int node : nodeList) {
            temp.add(node);
            dfs(graph, graph[node], temp, endNode);
            temp.remove(temp.size() - 1);
        }
    }
}

                


